### PR TITLE
Added methods to JavaOSGiTest to retrieve multiple instances of a service.

### PR DESCRIPTION
**Pull request by ViToni**
Signed-off-by: Victor Toni <victor.toni@gmail.com>

### CODE REVIEW

**Comment by maggu2810** [comment]
It is a little bit API breaking but I assume no one will be affected.

You changed a `protected` method that has been allowed to return null that it must return a non-null value. Sub-classes that have overwritten that function violates that constrain now if they return null.

**Comment by maggu2810** [comment]
Can be merged (IMHO) as soon as Travis CI finished / succeeded.

File: bundles/test/org.eclipse.smarthome.test/src/main/java/org/eclipse/smarthome/test/java/JavaOSGiTest.java [hunk 1]
```diff
@@ -101,39 +104,58 @@ public void bindBundleContext() {
...
+
+    /**
+     * Get the single instance of an OSGi service or throw an {@link AssertionError} if multiple instances were found.
+     *
+     * @param clazz under which the OSGi service is registered
+     * @param filteredServices List of found services
+     * @return OSGi service or null if no service was found for the given class
+     * @throws AssertionError if more than one instance of the service is found
+     */
+    private <I, T extends I> @Nullable I getSingleServiceInstance(Class<T> clazz, final List<I> filteredServices) {
```
[the remark this answers is not in the document]
Comment (ViToni) [reply bundles/test/org.eclipse.smarthome.test/src/main/java/org/eclipse/smarthome/test/java/JavaOSGiTest.java]: Fixed.

File: bundles/test/org.eclipse.smarthome.test/src/main/java/org/eclipse/smarthome/test/java/JavaOSGiTest.java [hunk 2]
```diff
@@ -149,16 +171,41 @@ public void bindBundleContext() {
...
      * @param implementationClass the implementation class
-     * @return OSGi service or null if no service can be found for the given class
+     * @return OSGi service or null if no service can be found for the given classes
+     * @throws AssertionError if more than one instance of the service is found
      */
-    protected <T, I extends T> @Nullable I getService(Class<T> clazz, Class<I> implementationClass) {
-        @SuppressWarnings("unchecked")
-        final I service = (I) getService(clazz, srvRef -> implementationClass.isInstance(unrefService(srvRef)));
-        return service;
+    protected <I, T extends I> @Nullable I getService(Class<T> clazz, Class<I> implementationClass) {
```
[the remark this answers is not in the document]
Comment (ViToni) [reply bundles/test/org.eclipse.smarthome.test/src/main/java/org/eclipse/smarthome/test/java/JavaOSGiTest.java]: Fixed.

File: bundles/test/org.eclipse.smarthome.test/src/main/java/org/eclipse/smarthome/test/java/JavaOSGiTest.java [hunk 3]
```diff
@@ -149,16 +171,41 @@ public void bindBundleContext() {
...
+    }
+
+    /**
+     * Get all OSGi services for the given service class and the given implementation class.
+     *
+     * @param clazz class under which the OSGi services are registered
+     * @param implementationClass the implementation class of the services
+     * @return List of OSGi service or empty List if no matching services can be found for the given classes
+     */
+    protected <T, I> List<I> getServices(Class<T> clazz, Class<I> implementationClass) {
```
[the remark this answers is not in the document]
Comment (ViToni) [reply bundles/test/org.eclipse.smarthome.test/src/main/java/org/eclipse/smarthome/test/java/JavaOSGiTest.java]: Fixed.

**Comment by ViToni** [comment]
> It is a little bit API breaking but I assume no one will be affected.
> 
> You changed a `protected` method that has been allowed to return null that it must return a non-null value. Sub-classes that have overwritten that function violates that constrain now if they return null.
@maggu2810 
Could you point to the method mentioned? The basic idea is to have `getService_s_` which always return non-null `List`s (with non-null) service instances (if any).

If only one (or no) instance is to be expected the already existing `getService` should be used.
The implementation has been changed to use `getService_s_` internally without changing the current behaviour.

**Comment by maggu2810** [comment]
My comment has been about

```Java
protected <T> @Nullable T getService(Class<T> clazz, Predicate<ServiceReference<T>> filter)
```

and your change that the return value is not nullable anymore.

The method is not `final` and so a subclass could overwrite that method and return `null` in some cases.

I don't think it has been done by anyone, but the compilation of a subclass violates the nullness annotation now if it returns null.

**Comment by maggu2810** [comment]
> Could you double check please @maggu2810.

Too late :wink:

**Comment by maggu2810** [comment]
I am slowly but surely going crazy...

All fine, the side-by-side diff confused me.

**Comment by htreu** [comment]
All good, me too. I had to look 10 times :-)

**Pull request merged**